fasm/__init__.py

```diff
@@ -46,122 +60,134 @@ class ValueFormat(enum.Enum):
-    if not fasm_model:
-        return
-
-    for fasm_line in fasm_model.lines:
-        set_feature = None
-        annotations = None
-        comment = None
-
-        if fasm_line.set_feature:
-            set_feature = set_feature_model_to_tuple(fasm_line.set_feature)
-
-        if fasm_line.annotations:
-            annotations = tuple(
-                Annotation(
-                    name=annotation.name,
-                    value=annotation.value if annotation.value else '')
-                for annotation in fasm_line.annotations.annotations)
-
-        if fasm_line.comment:
-            comment = fasm_line.comment.comment
-
-        yield FasmLine(
-            set_feature=set_feature,
-            annotations=annotations,
-            comment=comment,
-        )
-
+        feature = feature,
+        start = start,
+        end = end,
+        value = value,
+        value_format = value_format
+    ), i+length
+
+# Decode an annotation: x = "y"
+def fasm_annotation_from_bytes(data, i):
+    length, i = get_header('a', data, i)
+    if not length:
+        return None, i
+
+    name, p = tagged_string_from_bytes('.', data, i)
+    value, p = tagged_string_from_bytes('=', data, p)
+    return Annotation(
+        name = name,
+        value = value
+    ), i+length
+
+# Decode a set of annotations: { ... }
+def fasm_annotations_from_bytes(data, i):
+    length, i = get_header('{', data, i)
+    if not length:
+        return None, i
+
+    annotations = []
+    annotation, p = fasm_annotation_from_bytes(data, i)
+    while annotation:
+        annotations.append(annotation)
+        annotation, p = fasm_annotation_from_bytes(data, p)
+
+    return annotations, i+length
+
+# Decode an entire FASM line
+def fasm_line_from_bytes(data, i):
+    length, i = get_header('l', data, i)
+    if not length:
+        return None, i
+
+    set_feature, p = fasm_set_feature_from_bytes(data, i)
+    annotations, p = fasm_annotations_from_bytes(data, p)
+    comment, p = tagged_string_from_bytes('#', data, p)
+
+    return FasmLine(
+        set_feature = set_feature,
+        annotations = annotations,
+        comment = comment
+    ), i+length
+
+
+def parse_fasm_data(data):
+    """ Parse FASM string, returning list of FasmLine named tuples."""
+    lines = []
+    line, p = fasm_line_from_bytes(data, 0)
+    while line:
+        lines.append(line)
+        line, p = fasm_line_from_bytes(data, p)
+    return lines
 
 def parse_fasm_string(s):
     """ Parse FASM string, returning list of FasmLine named tuples."""
-    return fasm_model_to_tuple(get_fasm_metamodel().model_from_str(s))
-
+    return parse_fasm_data(parse_fasm.from_string(bytes(s, 'ascii'), 0))
 
 def parse_fasm_filename(filename):
     """ Parse FASM file, returning list of FasmLine named tuples."""
-    return fasm_model_to_tuple(get_fasm_metamodel().model_from_file(filename))
+    return parse_fasm_data(parse_fasm.from_file(bytes(filename, 'ascii'), 0))
```

Comment:
I recommend trying if using Python `buffer`s help here: https://docs.python.org/3/c-api/buffer.html  They provide cheap slicing, and might provide a nice speed boost if the `bytes` slicing is causing excessive copies.

Reply:
The problem I had was that the slice doesn't know where it ends, so I can't use `data[i:]` like an iterator in C++. That was my initial approach, but it gave an error that seems specific to ctypes. So I'd have to track how many bytes are left in a line even though it's not needed.

Just tracking the offset wasn't too painful.

Reply:
The reason to use the memoryview is to avoid copying the data on slice.  If this is a large lift, fine.  But I think there is likely some performance to be gained by avoiding copies on slice.

Also I believe the memoryview can track the length of the underlying data, see the documentation: https://docs.python.org/dev/library/stdtypes.html#memoryview

```
len(view) is equal to the length of tolist. If view.ndim = 0, the length is 1. If view.ndim = 1, the length is equal to the number of elements in the view. 
```

Also see the examples:
```
A memoryview supports slicing and indexing to expose its data. One-dimensional slicing will result in a subview:

>>>
>>> v = memoryview(b'abcefg')
>>> v[1]
98
>>> v[-1]
103
>>> v[1:4]
<memory at 0x7f3ddc9f4350>
>>> bytes(v[1:4])
b'bce'
```

Reply:
These strings are just being copied. They are encoded in ASCII (1-byte chars) from Python's native string format, which I believe uses something like `wchar_t`. I looked at `memoryview`, but I don't copy the data except to populate the model, which is unavoidable.